web/src/main/webapp/heatmap3d.ts

```diff
@@ -0,0 +1,804 @@
+    protected chartsDiv: any;
+    protected legendDiv: any;
+    protected summary: any;
+    protected pointWidth: number;
+    protected pointHeight: number;
+    private scrollBar: ScrollBar;
+
+    protected currentData: {
+        xData: AxisData;
+        yData: AxisData;
+        zData: AxisData;
+        missingData: number;
+        data: number[][][];
+        xPoints: number;
+        yPoints: number;
+        zPoints: number;
+    };
+    private overlayLegend: HeatMapLegend;
+    private logScale: boolean;
+
+    constructor(remoteObjectId: string, protected tableSchema: Schema, page: FullPage) {
+        super(remoteObjectId, page);
+        this.topLevel = document.createElement("div");
+        this.topLevel.classList.add("chart");
+        this.topLevel.classList.add("heatMapArray");
+        this.logScale = false;
+        let menu = new TopMenu( [
+            { text: "View", subMenu: new TopSubMenu([
+                { text: "refresh", action: () => { this.refresh(); } },
+                { text: "table", action: () => { this.showTable(); } },
+            ]) },
+            {
+                text: "Combine", subMenu: combineMenu(this)
+            }
+        ]);
+        this.topLevel.appendChild(menu.getHTMLRepresentation());
+        this.topLevel.tabIndex = 1;
+        let chartsAndScrollbarDiv = document.createElement("div");
+
+        this.chartsDiv = d3.select(chartsAndScrollbarDiv)
+            .append("div")
+            .attr("class", "array");
+        this.overlayLegend = new HeatMapLegend();
+        this.topLevel.appendChild(this.overlayLegend.getHTMLRepresentation());
+
+        this.scrollBar = new ScrollBar(this);
+
+        chartsAndScrollbarDiv.style.flexDirection = "row";
+        chartsAndScrollbarDiv.style.display = "flex";
+        chartsAndScrollbarDiv.style.flexWrap = "nowrap";
+        chartsAndScrollbarDiv.style.justifyContent = "flex-start";
+        chartsAndScrollbarDiv.style.alignItems = "stretch";
+        this.topLevel.appendChild(chartsAndScrollbarDiv);
+        chartsAndScrollbarDiv.appendChild(this.scrollBar.getHTMLRepresentation());
+
+    }
+
+    // combine two views according to some operation
+    combine(how: CombineOperators): void {
+        let r = SelectedObject.current.getSelected();
+        if (r == null) {
+            this.page.reportError("No view selected");
+            return;
+        }
+
+        let rr = this.createRpcRequest("zip", r.remoteObjectId);
+        let renderer = (page: FullPage, operation: ICancellable) => {
+            return new Make2DHistogram(
+                page, operation,
+                [this.currentData.xData.description, this.currentData.yData.description],
+                this.tableSchema, true);
+        };
+        rr.invoke(new ZipReceiver(this.getPage(), rr, how, renderer));
+    }
+
+    public static logScale(max: number) {
+        return max > 20;
+    }
+
+    public static scale(max: number, logScale: boolean) {
```

Comment:
This stuff is also in common with HeatMapView, maybe you can refactor it into a common base class
HeatMapViewBase